core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -1524,136 +1524,89 @@ public static int multiply(int b0, int b1) {
-  public static @Nullable Long safeMultiply(long b0, int b1) {
-    BigDecimal ans = multiply(BigDecimal.valueOf(b0), BigDecimal.valueOf(b1));
-    boolean overflow =
-        isOverflow(ans, BigDecimal.valueOf(Long.MAX_VALUE), BigDecimal.valueOf(Long.MIN_VALUE));
-    return overflow ? null : ans.longValue();
+  /** SQL <code>SAFE_MULTIPLY</code> function applied to long and double values. */
+  public static @Nullable Double safeMultiply(long b0, double b1) {
+    double ans = b0 * b1;
+    boolean isInfinite = Double.isInfinite(b1);
+    return doubleOverflow(ans) || isInfinite ? ans : null;
   }
 
-  /** SQL <code>SAFE_MULTIPLY</code> function applied to long and double values. */
-  public static @Nullable Long safeMultiply(long b0, double b1) {
-    BigDecimal ans = multiply(BigDecimal.valueOf(b0), BigDecimal.valueOf(b1));
-    boolean overflow =
-        isOverflow(ans, BigDecimal.valueOf(Long.MAX_VALUE), BigDecimal.valueOf(Long.MIN_VALUE));
-    return overflow ? null : ans.longValue();
+  /** SQL <code>SAFE_MULTIPLY</code> function applied to double and BigDecimal values. */
+  public static @Nullable Double safeMultiply(double b0, BigDecimal b1) {
+    double ans = b0 * b1.doubleValue();
+    boolean isInfinite = Double.isInfinite(b0);
+    return doubleOverflow(ans) || isInfinite ? ans : null;
   }
 
-  /** SQL <code>SAFE_MULTIPLY</code> function applied to long values. */
-  public static @Nullable Long safeMultiply(long b0, long b1) {
-    BigDecimal ans = multiply(BigDecimal.valueOf(b0), BigDecimal.valueOf(b1));
-    boolean overflow =
-        isOverflow(ans, BigDecimal.valueOf(Long.MAX_VALUE), BigDecimal.valueOf(Long.MIN_VALUE));
-    return overflow ? null : ans.longValue();
+  /** SQL <code>SAFE_MULTIPLY</code> function applied to BigDecimal and double values. */
+  public static @Nullable Double safeMultiply(BigDecimal b0, double b1) {
+    double ans = b0.doubleValue() * b1;
+    boolean isInfinite = Double.isInfinite(b1);
+    return doubleOverflow(ans) || isInfinite ? ans : null;
   }
 
-  /** SQL <code>SAFE_MULTIPLY</code> function applied to long and BigDecimal values. */
-  public static @Nullable BigDecimal safeMultiply(long b0, BigDecimal b1) {
-    BigDecimal ans = multiply(BigDecimal.valueOf(b0), b1);
-    boolean overflow =
-        isOverflow(ans, BigDecimal.valueOf(Float.MAX_VALUE), BigDecimal.valueOf(-Float.MAX_VALUE));
-    return overflow ? null : ans;
+  /** SQL <code>SAFE_MULTIPLY</code> function applied to double values. */
+  public static @Nullable Double safeMultiply(double b0, double b1) {
+    double ans = b0 * b1;
+    boolean isInfinite = Double.isInfinite(b0) || Double.isInfinite(b1);
+    return doubleOverflow(ans) || isInfinite ? ans : null;
+  }
+
+  /** Helper for the safe arithmetic functions to determine
+   * overflow for a BigDecimal value. According to BigQuery, BigDecimal
+   * overflow occurs if the precision is greater than 76 or the precision
+   * is greater than 38. */
```

Comment:
The implementation of this function would return true if b.scale is 2 and b.precision is 5. That doesn't seem right.

Reply:
The behavior is correct, the naming is just a bit confusing. I changed it to `safeDouble` and `safeDecimal` which return `TRUE` if there is not overflow and `FALSE` if there is. Let me know if you agree with this.

Reply:
Is that sentence correct?
`if the precision is greater than 76 or the precision is greater than 38.`

Should one of them be scale?

Reply:
Oops, good catch. Fixed now.